metronome/hotstuff/service/src/io/iohk/metronome/hotstuff/service/RemoteConnectionManager.scala

```diff
@@ -0,0 +1,458 @@
+  ): F[Either[ConnectionFailure[K], ConnectionSuccess[F, K, M]]] = {
+    encryptedConnectionProvider
+      .connectTo(connectionRequest.key, connectionRequest.address)
+      .redeemWith(
+        e => Concurrent[F].delay(Left(ConnectionFailure(connectionRequest, e))),
+        connection => Concurrent[F].delay(Right(ConnectionSuccess(connection)))
+      )
+  }
+
+  case class RetryConfig(
+      initialDelay: FiniteDuration,
+      backOffFactor: Long,
+      maxDelay: FiniteDuration
+  )
+
+  object RetryConfig {
+    import scala.concurrent.duration._
+    def default: RetryConfig = {
+      RetryConfig(500.milliseconds, 2, 30.seconds)
+    }
+  }
+
+  private def retryConnection[F[_]: Timer: Concurrent, K](
+      config: RetryConfig,
+      connectionFailure: ConnectionFailure[K]
+  ): F[OutGoingConnectionRequest[K]] = {
+    // TODO add error logging
+    val updatedFailureCount =
+      connectionFailure.connectionRequest.numberOfFailures + 1
+    val exponentialBackoff =
+      math.pow(config.backOffFactor.toDouble, updatedFailureCount).toLong
+    val newDelay =
+      (config.initialDelay * exponentialBackoff).min(config.maxDelay)
+
+    Timer[F]
+      .sleep(newDelay)
+      .map(_ =>
+        connectionFailure.connectionRequest
+          .copy(numberOfFailures = updatedFailureCount)
+      )
+  }
+
+  /** Connections are acquired in linear fashion i.e there can be at most one concurrent call to remote peer.
+    * In case of failure each connection will be retried infinite number of times with exponential backoff between
+    * each call.
+    */
+  private def acquireConnections[F[
+      _
+  ]: Concurrent: TaskLift: TaskLike: Timer, K: Codec, M: Codec](
+      encryptedConnectionProvider: EncryptedConnectionProvider[F, K, M],
+      connectionsToAcquire: ConcurrentQueue[F, OutGoingConnectionRequest[K]],
+      connectionsRegister: ConnectionsRegister[F, K, M],
+      connectionsQueue: ConcurrentQueue[F, EncryptedConnection[F, K, M]],
+      retryConfig: RetryConfig
+  ): F[Unit] = {
+
+    /** Observable is used here as streaming primitive as it has richer api than Iterant and have mapParallelUnorderedF
+      * combinator, which makes it possible to have multiple concurrent retry timers, which are cancelled when whole
+      * outer stream is cancelled
+      */
+    Observable
+      .repeatEvalF(connectionsToAcquire.poll)
+      .mapEvalF { connectionToAcquire =>
+        connectTo(encryptedConnectionProvider, connectionToAcquire)
+      }
+      .mapParallelUnorderedF(Integer.MAX_VALUE) {
+        case Left(failure) =>
+          retryConnection(retryConfig, failure).flatMap(updatedRequest =>
+            connectionsToAcquire.offer(updatedRequest)
+          )
+        case Right(connection) =>
+          connectionsRegister
+            .registerConnection(connection.encryptedConnection)
+            .flatMap(_ =>
+              connectionsQueue.offer(connection.encryptedConnection)
+            )
```

Comment:
so in case of have disjointed design (i.e manager known topology before hand with incoming and outgoing connection) this just would not happen, but after our discussion that manager should kinda resolve who will be connected to who by iteslf some additinal checks will need to be added here.